src/els_text_edit.erl

```diff
@@ -0,0 +1,71 @@
+-module(els_text_edit).
+
+-export([ diff_files/2
+        ]).
+
+-include("erlang_ls.hrl").
+
+%%==============================================================================
+%% File and module functions
+%%==============================================================================
+
+%% @doc Construct a [text_edit()] from the diff between two files.
+%%
+%% Generate a list of text edits to transform `SourceFile' into
+%% `DestinationFile'
+-spec diff_files(els_uri:path(), els_uri:path()) -> [text_edit()].
+diff_files(SourceFile, DestinationFile) ->
+    Diffs = tdiff:diff_files(SourceFile, DestinationFile),
+    Patch = make_text_edits(Diffs),
+    Patch.
+
+
+%%==============================================================================
+%% Internal functions
+%%==============================================================================
+
+%% A text_edit() has a range and a blob of text.
+%%
+%% insert:
+%%    start and end of range are the same, specifying where to include
+%% delete:
+%%    start and end of range define region to be removed, blank text.
+%% change:
+%%    start and end are original range, text is replacement text
+
+-type diff() :: {del, [string()]} | {eq, [string()]} | {ins, [string()]}.
+
+-spec make_text_edits([diff()]) -> [text_edit()].
+make_text_edits(Diffs) ->
+    make_text_edits(Diffs, 0, []).
+
+-spec make_text_edits([diff()], number(), [text_edit()]) -> [text_edit()].
+make_text_edits([{eq, Data}|T], Line, Acc) ->
+    make_text_edits(T, Line + length(Data), Acc);
+
+make_text_edits([{del, Del}, {ins, Ins}|T], Line, Acc) ->
+    Len = length(Del),
+    Pos1 = #{ line => Line,       character => 0 },
+    Pos2 = #{ line => Line + Len, character => 0 },
+    Edit = #{ range => #{ start => Pos1, 'end' => Pos2 }
+            , newText => list_to_binary(lists:concat(Ins))
```

Comment:
This might now work correctly with utf8 data. Is `Ins` a list of bytes or code-points? `list_to_binary` generally fails on elements `> 255`. Probably `unicode:characters_to_binary` would be better in here, and then `lists:concat` is not needed.

Reply:
`lists:concat` is used to join up the lines again. Agree on using `unicode:characters_to_binary`.